pkg/diagnostics/networkpod/pod.go

```diff
@@ -0,0 +1,288 @@
+	globalPods := []kapi.Pod{}
+	nonGlobalPods := []kapi.Pod{}
+
+	for _, pod := range pods {
+		if len(vmap) > 0 && vmap[pod.Namespace] == api.GlobalVNID {
+			globalPods = append(globalPods, pod)
+		} else {
+			nonGlobalPods = append(nonGlobalPods, pod)
+		}
+	}
+	return globalPods, nonGlobalPods
+}
+
+func getOpenShiftNetworkPlugin(osClient *osclient.Client) (string, bool, error) {
+	cn, err := osClient.ClusterNetwork().Get(api.ClusterNetworkDefault)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return "", false, nil
+		}
+		return "", false, err
+	}
+	return cn.PluginName, sdnplugin.IsOpenShiftNetworkPlugin(cn.PluginName), nil
+}
+
+func getLocalAndNonLocalPods(kubeClient *kclient.Client) ([]kapi.Pod, []kapi.Pod, error) {
+	pods, err := getPods(kubeClient)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	localIP, err := getLocalNodeIP(kubeClient)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	localPods := []kapi.Pod{}
+	nonlocalPods := []kapi.Pod{}
+	for _, pod := range pods {
+		if pod.Status.HostIP == localIP {
+			localPods = append(localPods, pod)
+		} else {
+			nonlocalPods = append(nonlocalPods, pod)
+		}
+	}
+	return localPods, nonlocalPods, nil
+}
+
+func getLocalNodeIP(kubeClient *kclient.Client) (string, error) {
+	nodeList, err := kubeClient.Nodes().List(kapi.ListOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	_, hostIPs, err := netutils.GetHostIPNetworks(nil)
+	if err != nil {
+		return "", err
+	}
+	for _, node := range nodeList.Items {
+		if len(node.Status.Addresses) == 0 {
+			continue
+		}
+		for _, ip := range hostIPs {
+			for _, addr := range node.Status.Addresses {
+				if addr.Type == kapi.NodeInternalIP && ip.String() == addr.Address {
+					return addr.Address, nil
+				}
+			}
+		}
+	}
+	return "", fmt.Errorf("unable to find local node IP")
+}
+
+func getPods(kubeClient *kclient.Client) ([]kapi.Pod, error) {
+	podList, err := kubeClient.Pods(kapi.NamespaceAll).List(kapi.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	filtered_pods := []kapi.Pod{}
+	for _, pod := range podList.Items {
+		// Skip pods that are not running
+		if pod.Status.Phase != kapi.PodRunning {
+			continue
+		}
+
+		// Skip pods with hostNetwork enabled
+		if pod.Spec.SecurityContext.HostNetwork {
+			continue
+		}
+		filtered_pods = append(filtered_pods, pod)
+	}
+	return filtered_pods, nil
+}
+
+func checkDifferentNodePodToPodConnection(localGlobalPods, localNonGlobalPods, nonlocalGlobalPods, nonlocalNonGlobalPods []kapi.Pod, vmap map[string]uint32, r types.DiagnosticResult) {
+	// Applicable to flat and multitenant networks
+	if len(localNonGlobalPods) > 0 && len(nonlocalNonGlobalPods) > 0 {
+		checkPodToPodConnection(localNonGlobalPods[0], nonlocalNonGlobalPods[0], vmap, r)
+	} else {
+		r.Warn("DPodNet1005", nil, fmt.Sprintf("Skipping pod connectivity test for non-global projects on different nodes. Reason: Couldn't find 2 non-global pods."))
+	}
+	if len(vmap) == 0 {
+		return
+	}
```

Comment:
This seems fragile; it arbitrarily splits the function into two unequal portions. (Eg, if you want to add another test to this function in the future, you can't just append it to the end of the function. Depending on what kind of test it is, you either *must* add it before this line, or *must* add it after this line.)

Doing "if len(vmap) != 0 { /* multi-tenant-only tests in here... */ }" would be a little clearer. Or add the vmap test to each individual multi-tenant-only test. Or split the multi-tenant-only tests into a separate function.